september-update.md

```diff
@@ -0,0 +1,125 @@
+to reason about, leads to surprising outcomes vs. f-strings, and challenges static analysis. The feeling: choosing "
+lazy" should be opt-in, and the coder should make that choice e.g. via lambda wrapping the interpolation.
+
+We now use eager evaluation of interpolation, storing the result on `Interpolation.value`. We still believe lazy
+evaluation has a place, but will defer (hah!) that to another PEP. We also believe there are ways to keep some of the
+deferred approach, which we plan to briefly outline. Finally, we envision DSLs with intermediate representations (i.e.
+ASTs) which lazily-render to a string as a later step.
+
+## Template strings
+
+The first version followed JS by using tag functions as prefixes to f-string-like strings. For example:
+`html"<div>Hello {name}</div>"`. This met serious resistance:
+
+- Very hard to teach/discover
+- No dotted names
+- Namespace pollution
+
+Our approach combined two steps into one: converting the string into a standard data structure, then running
+a DSL function with that data. We now plan to split those steps: `html(t"<div>Hello {name}</div>")`. Template string ->
+template function.
+
+We now propose a single `t` prefix, as suggested in the thread. It generates a `Template` which can be stored as a
+variable, passed to a callable -- all regular normal Python stuff.
+
+## `Template` and `Interpolation`
+
+The previous version passed a sequence of `args` to the tag function. An `arg` was either a `Decoded` or
+`Interpolation`.
+
+Instead, we will gather all the information into a `Template`:
+
+```python
+@runtime_checkable
+class Template(Protocol):
+    args: tuple[Decoded | Interpolation, ...]
+    source: str
+```
+
+`Template` is a runtime-checkable protocol with a built-in implementation of `TemplateConcrete`. But template functions
+can be called with custom classes, e.g. for testing or alternate implementations. `Template.source` has the original
+text of the template string.
+
+The other change: the `Interpolation.getvalue()` callable is replaced by `Interpolation.value`. It contains the
+eagerly-evaluated interpolation result. We can thus drop the work done for annotation scopes.
+
+## DSL typing
+
+The discussion showed strong interest in the DSL nature of this. That is, a way to signal the language to be used in a
+template string.
+
+`Template` and `TemplateConcrete` don't contain any information about the DSL. We believe this is a good thing. But we
+do plan a normative section on how another PEP might do this. In fact, we want this PEP's goal to remain focused on DSLs
+and developer experience.
+
+We plan to *suggest* (again: normative) putting the language information into type information. This makes it available
+for static analysis. For example, signaling an HTML DSL:
+
+```python
+@runtime_checkable
+class Template(Protocol):
+    args: tuple[Decoded | Interpolation, ...]
+    source: str
+
+
+@runtime_checkable
```

Comment:
By the way, it's probably worth pointing out that with this approach, there's no static distinction between `Template` and `HTMLTemplate`. The extra annotations on `HTMLTemplate.source` don't matter.

And so type checkers won't flag this:

```python
def evil(t: Template) -> HTMLTemplate:
    return t  # *is* this Template really an HTMLTemplate? Alas, the type system cannot tell.
```

At runtime, it will also always be the case that `isinstance(foo, Template) == isinstance(foo, HTMLTemplate)`.

The only way to distinguish the two at runtime is with `typing.get_type_hints(..., include_extras=True)`. That'll return different values for `Template` and `HTMLTemplate`.

Reply:
Damn, you're right! From the mypy docs: `Note that inheriting from an existing protocol does not automatically turn the subclass into a protocol`. Fixed.